toolbox/api/datagalaxy_api_modules.py

```diff
@@ -0,0 +1,121 @@
+        headers = {'Authorization': f"Bearer {self.token}"}
+        response = requests.get(f"{self.url}/dataProcessingItem", params=params, headers=headers)
+        code = response.status_code
+        body_json = response.json()
+        if code != 200:
+            raise Exception(body_json['error'])
+        result = []
+        result = result + body_json['results']
+        next_page = body_json["next_page"]
+        while next_page is not None:
+            headers = {'Authorization': f"Bearer {self.token}"}
+            response = requests.get(next_page, headers=headers)
+            body_json = response.json()
+            next_page = body_json["next_page"]
+            result = result + body_json['results']
+        return result
+
+    def bulk_upsert_tree(self, workspace_name: str, objects: list, tag_value: Optional[str]) -> int:
+        # Objects can be in pages, so one POST request per page
+        for page in objects:
+            # Existing entities are updated and non-existing ones are created.
+            bulktree = build_bulktree(page)
+
+            if tag_value is not None:
+                bulktree = prune_tree(bulktree, tag_value)
+
+            # If a parent usage has a technology, it is necessary to delete the "technologyCode" property in every children
+            # Otherwise the API returns an error. Only the parent can hold the "technologyCode" property
+            for tree in bulktree:
+                if 'children' in tree:
+                    for children in tree['children']:
+                        remove_technology_code(children)
+
+            version_id = self.workspace['defaultVersionId']
+            headers = {'Authorization': f"Bearer {self.token}"}
+            response = requests.post(f"{self.url}/{self.route}/bulktree/{version_id}", json=bulktree, headers=headers)
+            code = response.status_code
+            body_json = response.json()
+            if 200 <= code < 300:
+                logging.info(f'bulk_upsert_tree - {body_json}')
+            if 400 <= code < 500:
+                raise Exception(body_json['error'])
+
+        return 200
+
+    def delete_objects(self, workspace_name: str, ids: list) -> int:
+        if len(ids) < 1:
+            logging.warn(f'Nothing to delete on workspace "{workspace_name}" in module {self.module}, aborting.')
+            return 0
+        version_id = self.workspace['defaultVersionId']
+        headers = {'Authorization': f"Bearer {self.token}"}
+        response = requests.delete(f"{self.url}/{self.route}/bulk/{version_id}",
+                                   json=ids,
+                                   headers=headers)
+        code = response.status_code
+        body_json = response.json()
+        if code != 200:
+            raise Exception(body_json['error'])
+        logging.info(
+            f'delete_objects - {body_json['totalDeleted']} objects were deleted on workspace "{workspace_name}" in module {self.module}')
```

Comment:
Using our CI (Python 3.9) we have:

```
./toolbox/api/datagalaxy_api_modules.py:120:44: E999 SyntaxError: f-string: unmatched '['
```

If it's working on our side, I guess you are using python 3.12, which is lifting some restrictions on the usage of `f-strings`

See:
https://docs.python.org/3.12/whatsnew/3.12.html#pep-701-syntactic-formalization-of-f-strings

Reply:
Nice catch! Fixed